pkg/datapath/linux/devices.go

```diff
@@ -501,12 +503,7 @@ func findK8SNodeIPLink() (netlink.Link, error) {
 // supportL3Dev returns true if the kernel is new enough to support BPF host routing of
 // packets coming from L3 devices using bpf_skb_redirect_peer.
 func supportL3Dev() bool {
-	probesManager := probes.NewProbeManager()
-	if h := probesManager.GetHelpers("sched_cls"); h != nil {
-		_, found := h["bpf_skb_change_head"]
-		return found
-	}
-	return false
+	return probes.HaveProgramHelper(ebpf.SchedCLS, asm.FnSkbChangeHead) == nil
```

Comment:
Incidentally, it seems like a curious choice to expose the functions via the `asm` package rather than just in the ebpf package, since the asm defines the BPF_CALL instruction but the interpretation of the arguments (ie numbers here) is up to the eBPF runtime. That is to say, depending on which platform we target, the actual meaning of the number 5 which is currently mapped to some specific function in the asm package could be different. But that's probably more feedback for ensuring the ebpf library is cross-platform rather than this specific PR. (Though that said, if we change this in the upstream library then obviously this PR would also need updating).

Reply:
^^ I'm not quite sure what this means in the end. Maybe creating a `github.com/cilium/ebpf/linux` package specifically to define these constants for the Linux eBPF runtime?

I'll cc @dthaler here in case he has thoughts / opinions on this, given his background on the eBPF for Windows project.

Reply:
> Incidentally, it seems like a curious choice to expose the functions via the asm package rather than just in the ebpf package ...

That sounds like a strawman argument to me. The `asm` package is where the ISA description lives and is used by the instruction decoder. When it was conceived, it made sense to put these constants here, since Linux had the only real eBPF implementation at the time (ignoring ubpf for a second), and why would anyone ever want to write another runtime with different helpers? :wink:

Since helpers are part of the ABI and we were only targeting Linux, they were implicitly considered part of the ISA. Until the `features` package came around, helper type constants weren't taken by any APIs outside of the `asm` package, so they were really values you'd only use when constructing bytecode from Go. It may look a bit off-color, but there's still not much use for them elsewhere in the API, so IMO they're fine where they are.

We could consider putting these behind build tags, but the existing ones will need to remain available in `asm` for compatibility. I don't see a real need to change this until someone comes along to add Windows support, though.

Reply:
I raise this question specifically because of the active discussions around cross-platform support going on at the moment in the context of the eBPF for Windows project. Cross-platform eBPF in general was discussed as recently as two weeks ago as part of the eBPF community call ([notes](https://lore.kernel.org/bpf/CH2PR21MB14649A069BDEACFBC440C623A3649@CH2PR21MB1464.namprd21.prod.outlook.com/T/#u)). One of the points highlighted from recent discussions is that the helper codes for that instruction are not guaranteed to be consistent between platforms. When you look at the current helpers, some of those helpers are very Linux specific, so it will not make sense for other platforms to support the same numbers for those helpers. From what I understand, the current thinking from other runtimes is that they will not ensure ABI compatibility with the Linux runtime for the `BPF_CALL` instruction.

From your explanation, I think there is a completely reasonable line of thinking that leads to the current layout of these. That said, if the question is as simple as which package these symbols live in, I think it's worth at least considering how a future of cilium/ebpf against other runtimes might work so we can minimize any library change impacts in future. Maybe the future answer to this should be as simple as setting some Go build flags when using the eBPF library in order to target specific runtimes. In that case, no further changes would be necessary. I floated the idea above that these ABI symbols could be declared in a separate package, that's also a possibility but I don't feel strongly that this should be the case.